### ISSUE TITLE
integrate fuzzer into our workflow

### ISSUE THREAD

**Issue by ailin-nemui** [irssi/irssi#547]

**Comment by josephbisch** [comment]
Is this something you are looking for help with? Do you mean that you are looking for someone to contribute instructions on how to fuzz irssi with say afl for example, or do you just mean that the irssi developers should run a fuzzer themselves?

**Comment by ahf** [comment]
Yes, we would love contributions in this area.

We're interested in both having people run their fuzzers against irssi and reporting issues they may find. Hanno Bock already reported one bug with our config parser and I just saw that you reported a similar one.

We're also interested in finding a way where we can automate the execution in a more systematic way.

**Comment by josephbisch** [comment]
Yes, automating the execution would be nice. I'll try looking into that.

Should I ignore crashes like the following? The message does recognize that the config may be corrupt, so maybe there is a certain grammar for the config file that afl should be made aware of through an afl dictionary. Or are those crashes of interest?

```
** (irssi:30174): CRITICAL **: Expected block node at `..(null)/c0000000' was of scalar type. Corrupt config?
Trace/breakpoint trap (core dumped)
```

**Comment by ahf** [comment]
@ailin-nemui might have a say to that :-)

**Comment by dequis** [comment]
That seems to be because of 1e4f7e63249b0300cac27e725e14a0134260a387 which turns `g_critical()` warnings into errors during initialization only. I suggest reverting that commit for fuzzing - it won't error out during /reload so maybe you can find interesting paths that way.

**Comment by josephbisch** [comment]
I found https://github.com/ozzyjohnson/docker-afl, but there isn't any license information there, so I would need to contact the author first to get a license put in the repo before using that as a starting point. Or it shouldn't be hard to create a dockerfile from scratch.

What do people think about using a Docker or LXC container versus using VMs (qemu or use vagrant to use virtualbox)? I'm leaning towards using a VM. We definitely need some sort of isolation from the network so that we don't spam irc channels with messages. I've been using Linux network namespaces for the fuzzing I have done so far, but VM provisioning scripts would be nice, since we don't know what setup people might have who want to run the fuzzer.

@hannob - What do you think? What do you use for The Fuzzing Project?

Also, I think there should be some way to easily view the fuzzing results. Maybe periodically run afl-tmin on the new results to get minimized testcases and have a web interface. The web interface would allow for viewing the testcase contents and the backtrace from gdb. Maybe have a copy-pastable bug report text with all the relevant information to easily submit a bug report. But a web interface like that is probably something that should go into afl itself. I'll see what the afl-users Google Group says about adding a web interface to afl versus making it a separate project.

**Comment by hannob** [comment]
I'm not using any containers or virtualization at all, I do all my fuzzing either locally for quick testing or on a dedicated server. I'm not familiar with the docker container.

**Comment by josephbisch** [comment]
@hannob - Ah okay, so then you must just have the dedicated server not connected to the internet when fuzzing software that may make internet connections or just avoid fuzzing those parts of such software? My concern is that the fuzzer will generate input to irssi that will cause messages to be sent to public irc channels if there isn't some sort of isolation from the network. And not everyone is using a dedicated machine to fuzz on, so physically disconnecting the machine isn't always an option. Thanks for letting us know your setup! Maybe irssi can have the option of running the fuzzer directly or using virtualization, so that people can choose what works best for their setup.

Having played with this some more, it seems like there needs to also be an irc server that runs locally and is accessible to the irssi being fuzzed. Otherwise it seems hard to test much of irssi's functionality without being able to connect to an irc server.

**Comment by josephbisch** [comment]
[oss-fuzz](https://github.com/google/oss-fuzz) should also be considered.

**Comment by josephbisch** [comment]
I'm going to go with afl for now and try to get something working. I'll also include a vagrantfile, so that the fuzzing can be done in a VM if the person fuzzing chooses. To be able to fuzz parts of irssi other than just the config file parser, I think an irc server needs to be run for us to fuzz against. Anyone have any opinions on the irc server? I could actually make the server used configurable.

Here is what I'm thinking:

1. Person fuzzing clones irssi git repo.
2. Person runs `./fuzzing-build` script that runs `./autogen.sh CC=afl-gcc` followed by `make`. Optionally, the fuzzing-build script takes a path to afl-gcc, so that it works even if the afl programs aren't on the path.
3. Optionally the person can install an irc server and run it locally. This is only needed if the fuzzing script that will be run connects to a server (i.e. config file parsing doesn't require this step).
4. There is an afl directory that is part of the irssi git repo. In it is some sort of config file that the user can use to specify various system-specific options (the memory flag for afl, the path to afl-fuzz, etc.) and a bunch of scripts that run afl. One such script may run something like `afl-fuzz -m 2G -i in-config -o out ../src/fe-text/irssi --config @@`. In the example, the in-config directory would contain the set of initial config files to start fuzzing with. There would be other in directories to cover the various types of input we need.
5. During fuzzing afl-tmin is run on the new files in the out directory.
6. Maybe there could be a Python or Golang based server component that can be used to view the outputs in a web browser.

**Comment by dequis** [comment]
I'm still not sure how it would make sense to have fuzzing as part of the workflow, and @ailin-nemui isn't explaining anything so...

Personally. what I do is patching irssi to move all network interaction to stdin/stdout.

Also, making the fuzzing results public immediately through a web UI sounds like a terrible idea. If there is something exploitable in there, it needs to be kept private. You know, coordinated disclosure et al.

**Comment by josephbisch** [comment]
I just meant a web UI as something like syzkaller has or Mozilla's FuzzManager. I was thinking just something that would run locally and not be made public.

**Comment by ailin-nemui** [comment]
My point in opening this issue was that some serious exploit were found with the most basic fuzzers [1] . I think we should have a way to easily run fuzzers, at least as part of the release process. However I don't know how to best go about it. 

[1] https://github.com/sandsmark/ircfuzz

**Comment by josephbisch** [comment]
Nobody seems to know how best to go about this. What if we do something general like add a make rule `make FUZZER=afl fuzz`? Forget about how to view/interpret results. Just let everyone decide themselves how they want to handle that part. I like the concept of ircfuzz too, but I can't find what the license is on it. Also I haven't found any crashes with ircfuzz, but I may have just not ran it long enough.

Have a file structure in the irssi source tree like:

```
-fuzzers
  -afl
    -build-irssi.sh
    -run-fuzzer.sh
  -otherfuzzer
    -build-irssi.sh
    -run-fuzzer.sh
```

I think there would also need to be a way to take input in on stdin or via a file (at least I do for afl, but the way ircfuzz works, it doesn't). We could have that built in via a preprocessor define if we are building for fuzzing purposes, otherwise it won't be built in.

What do people think about that? Just afl and ircfuzz (if ircfuzz is okay to use from a licensing perspective), at least to start? Is having a make rule like `make FUZZER=afl fuzz` the best way to do this from the perspective of the irssi release process?

**Comment by hannob** [comment]
I just checked out ircfuzz and it's interesting (although it doesn't find anything in irssi right away).

I've reached out to its author to ask if he is willing to put it under a free license.

**Comment by dequis** [comment]
Eh don't bother, it's the sort of fuzzer that is obsolete now that afl exists (Which makes the vuln fixed by 0.8.20 even more embarrassing)

**Comment by hannob** [comment]
FWIW the ircfuzz author has now published his code under 2-clause-BSD:
https://github.com/iljavs/ircfuzz/blob/master/ircfuzz.c

@dequis while AFL is certainly the more advanced fuzzing tool applying it to an irc client on this level isn't always trivial, as one would have to rewrite it in a way that it takes input from stdin and interpret it as network input. This would probably lead to more results, but I find a simple irc fuzzing tool still valuable.

**Comment by ailin-nemui** [comment]
if there are some simple code changes to irssi required we should discuss those. or otherwise evaluate how to pipe things together using socat2 and/or unix domain sockets which irssi already supports

**Comment by hannob** [comment]
Given the just released advisory with one bug I found via fuzzing I've blogged about it and published example scripts that can be used to fuzz some parts of irssi:

https://blog.fuzzing-project.org/55-Fuzzing-Irssi-with-Perl-Scripts.html

**Comment by josephbisch** [comment]
Thank you @hannob for linking your blog post and example scripts.

I should also link #610 here, so people are aware of it. It makes use of libfuzzer. Once that PR is accepted here, I plan to submit a PR to https://github.com/google/oss-fuzz that makes use of that fuzzing frontend.
